test/pylib/s3_proxy.py

```diff
@@ -0,0 +1,253 @@
+        self.send_response(code)
+        self.send_header('Content-Type', 'text/plain; charset=utf-8')
+        self.send_header('Connection', 'keep-alive')
+        req_uuid = str(uuid.uuid4())
+        response = f"""<?xml version="1.0" encoding="UTF-8"?>
+
+                                <Error>
+                                    <Code>{error_name}</Code>
+                                    <Message>Minio proxy injected error. Client should retry.</Message>
+                                    <RequestId>{req_uuid}</RequestId>
+                                    <HostId>Uuag1LuByRx9e6j5Onimru9pO4ZVKnJ2Qz7/C1NPcfTWAtRPfTaOFg==</HostId>
+                                </Error>""".encode('utf-8')
+        self.send_header('Content-Length', str(len(response)))
+        self.end_headers()
+        self.wfile.write(response)
+
+    def process_request(self):
+        try:
+            policy = self.get_policy()
+
+            if policy.error_count >= policy.max_errors:
+                policy.should_fail = False
+                policy.should_forward = True
+
+            response = Response()
+            body = None
+
+            content_length = self.headers['Content-Length']
+            if content_length:
+                body = self.rfile.read(int(content_length))
+
+            if policy.should_forward:
+                target_url = self.minio_uri + self.path
+                headers = {key: value for key, value in self.headers.items()}
+                response = requests.request(self.command, target_url, headers=headers, data=body)
+
+            if policy.should_fail:
+                policy.error_count += 1
+                self.respond_with_error()
+            else:
+                self.send_response(response.status_code)
+                for key, value in response.headers.items():
+                    if key.upper() != 'CONTENT-LENGTH':
+                        self.send_header(key, value)
+
+                if self.command == 'HEAD':
+                    self.send_header("Content-Length", response.headers['Content-Length'])
+                else:
+                    self.send_header("Content-Length", str(len(response.content)))
+                self.end_headers()
+                self.wfile.write(response.content)
+        except Exception as e:
+            print(e)
+
+    def do_GET(self):
+        self.process_request()
+
+    def do_POST(self):
+        self.process_request()
+
+    def do_PUT(self):
+        self.process_request()
+
+    def do_DELETE(self):
+        self.process_request()
+
+    def do_HEAD(self):
+        self.process_request()
+
+
+# Proxy server to setup `ThreadingHTTPServer` instance with custom request handler (see above), managing requests state
+# in the `self.req_states`, adding custom logger, etc. This server will be started automatically from `test.py` once the
+# pytest kicks in. In addition, it is possible just to start this server using another script - `start_s3_proxy.py` to
```

Comment:
> This server will be started automatically from `test.py` once the pytest kicks in.

i am confused. how does pytest kick in in the startup of the server?

Reply:
reworded